pep/005-mixins-are-bundles.md

```diff
@@ -0,0 +1,156 @@
+---
+title: "Mixins are Bundles"
+number: "005"
+status: "provisional"
+owner: "@carolynvs"
+authors:
+  - "@carolynvs"
+created: "2022-03-03"
+type: "feature"
+---
+
+## Abstract
+
+Let's take advantage of the distribution and security features of bundles for distributing and executing mixins.
+
+## Motivation
+
+Initial Proposal: https://github.com/getporter/porter/discussions/1499
+
+Currently mixins are binaries that are distributed by the porter client by downloading the mixin specified by the user and saving it to the porter home directory (PORTER_HOME).
+When a bundle is built, the mixin binaries located in PORTER_HOME are copied into the bundle's invocation image.
+This is a proposal to distribute the mixin binaries in a bundle instead of distributing the binaries directly.
+
+### Security
+
+___
+
+#### Verify Provenance
+
+> I want to be able to verify that the mixins in a bundle came from their stated source.
+
+Right now, we communicate the name of the mixins included in a bundle, and it is up to the user to decide if the bundle was built securely and actually contains the correct version of the stated mixins.
+A bundle author could use an arbitrary binary named after the mixin, and the end-user couldn't tell the difference.
+End-users should be able to inspect a bundle, verifying the mixins source and signatures, before running it.
+
+#### Secure Distribution
+
+> I want to distribute mixins using established, secure cloud infrastructure.
+
+By switching from manual binary hosting and distribution to bundles, OCI registries, and Notary, we can take advantage of Docker's security features, along with the features from the CNAB security specification.
+
+#### Isolate Sensitive Data
+> I want to limit which mixins can access bundle credentials. For example, only the az mixin should have access to my Azure credentials, not the exec mixin.
+
+In addition to distribution, there is a security concern around limiting access to sensitive data in a bundle.
+Every mixin in a bundle has access to every parameter, credential, and output.
+You may trust your Azure credentials with the az CLI, but exposing it to unrelated mixins is an unnecessary risk.
+Executing mixins inside a bundle allows us to limit the exposure of sensitive data on a need to know basis.
+For example, the az mixin would have access to your Azure credentials, but the exec mixin that can run arbitrary shell scripts would not.
+
+### Developer Experience
+___
+
+#### Reproducible Builds
+> I want my bundle to declare not only the mixin used, but its version and source.
+
+Bundle authors are responsible for tracking which version they use with each bundle and ensuring that version is installed before building the bundle, which is a major gap in our mixin functionality.
+Before a bundle can be built, it is the author must install the mixins used by the bundle, and guess at the proper version since the mixin version is not stored in the bundle's metadata.
```

Comment:
```suggestion
Before a bundle can be built, the author must install the mixins used by the bundle, and guess at the proper version since the mixin version is not stored in the bundle's metadata.
```